contracts/core/TokenAllocator.sol

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "@openzeppelin/contracts/proxy/utils/UUPSUpgradeable.sol";
+
+import "@solarity/solidity-lib/contracts-registry/AbstractDependant.sol";
+import "@solarity/solidity-lib/access-control/MultiOwnable.sol";
+
+import "../interfaces/core/ITokenAllocator.sol";
+import "../interfaces/core/IContractsRegistry.sol";
+
+contract TokenAllocator is ITokenAllocator, AbstractDependant, MultiOwnable, UUPSUpgradeable {
+    using MerkleProof for *;
+    using SafeERC20 for IERC20;
+    using EnumerableSet for *;
+
+    uint256 public lastAllocationId;
+    mapping(address => EnumerableSet.AddressSet) internal _tokensByAllocator;
+    mapping(address => EnumerableSet.AddressSet) internal _allocatorsByToken;
+    mapping(address => mapping(address => EnumerableSet.UintSet)) internal _allocations;
+    mapping(uint256 => AllocationData) internal _allocationInfos;
+
+    IPoolFactory internal _poolFactory;
+
+    event AllocationCreated(
+        uint256 id,
+        address allocator,
+        address token,
+        uint256 amount,
+        bytes32 merkleRoot,
+        string descriptionUrl
+    );
+    event AllocationClosed(uint256 id, address token, uint256 amountReturned);
+    event TokenClaimed(
+        address allocator,
+        address token,
+        bytes32 merkleRoot,
+        address user,
+        uint256 amount
+    );
+
+    modifier withCorrectId(uint256 id) {
+        require(id <= lastAllocationId, "TA: invalid allocation id");
+        _;
+    }
+
+    function __TokenAllocator_init() external initializer {
+        __MultiOwnable_init();
+    }
+
+    function setDependencies(
+        address contractsRegistry,
+        bytes memory data_
+    ) public override(AbstractDependant, ITokenAllocator) dependant {
+        IContractsRegistry registry = IContractsRegistry(contractsRegistry);
+
+        _poolFactory = IPoolFactory(registry.getPoolFactoryContract());
+    }
+
+    function createAllocation(
+        address allocator,
```

Comment:
Anyone can create "fake" allocations for any allocator. I see an attack vector here as trustworthy allocators can be spammed with junk.